chromium_src/base/test/launcher/test_launcher.cc

```diff
@@ -0,0 +1,83 @@
+/* Copyright (c) 2023 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include "base/test/launcher/test_launcher.h"
+
+#include "brave/base/test/launcher/teamcity_reporter.h"
+
+#define TestLauncher TestLauncher_ChromiumImpl
+#define AddTestResult(...) AddTestResult(OnTestResult(__VA_ARGS__));
+
+#include "src/base/test/launcher/test_launcher.cc"
+
+#undef TestLauncher
+#undef AddTestResult
+
+namespace base {
+
+TestLauncher::TestLauncher(TestLauncherDelegate* launcher_delegate,
+                           size_t parallel_jobs,
+                           size_t retry_limit)
+    : TestLauncher_ChromiumImpl(launcher_delegate, parallel_jobs, retry_limit) {
+  teamcity_reporter_ = TeamcityReporter::MaybeCreate();
+}
+
+TestLauncher::~TestLauncher() = default;
+
+// Upstream implementation of this method does roughly this:
+// 1. Print the test output if it has failed.
+// 2. Add test results via `results_tracker_.AddTestResult()`.
+// 3. Call exit(1) if a lot of test have failed.
+void TestLauncher::OnTestFinished(const TestResult& result) {
+  // The order of TC log calls is important here. First we want to let TC know a
+  // test is starting, then we call the original `OnTestFinished` which may
+  // print the test output on failure, so it will become a part of the
+  // TC-reported test. Finally we want to let TC know the test is finished so
+  // any other test launcher output should not be bound to the test.
+  if (teamcity_reporter_) {
+    teamcity_reporter_->OnTestStarted(result);
+  }
+
+  TestLauncher_ChromiumImpl::OnTestFinished(result);
```

Comment:
could we add a note, that `OnTestResult()` will be called here.

Reply:
added